README.md
tiny-level-ttl
==============

[![Build Status](https://travis-ci.org/ArtskydJ/tiny-level-ttl.svg?branch=master)](https://travis-ci.org/ArtskydJ/tiny-level-ttl)
[![Dependency Status](https://david-dm.org/artskydj/tiny-level-ttl.svg)](https://david-dm.org/artskydj/tiny-level-ttl)
[![devDependency Status](https://david-dm.org/artskydj/tiny-level-ttl/dev-status.svg)](https://david-dm.org/artskydj/tiny-level-ttl#info=devDependencies)

Enforces a [time to live (TTL)][wiki-ttl] on a [`node-levelup`][levelup] database.

- The ttl is specified in the constructor. (In [`node-level-ttl`][ttl] you can set the ttl in the constructor, and/or in a `db.put()` call.)
- Supports [`level-sublevel`][sublevel] 5 & 6, and [`level-spaces`][spaces] 1 & 2. (Defaults to [`level-spaces@2`][spaces].)
- Respects [`level-lock`][lock] locks.

Why use this instead of [`node-level-ttl`][ttl]? Because [`level-sublevel`][sublevel] and [`node-level-ttl`][ttl] conflict.

[Bug reproduction test code here.][bug-code]  
The bug was found using `level-sublevel@6` with `level-ttl@2`, and `level-sublevel@5` with `level-ttl 0.6`.

# api

```js
var ttl = require('tiny-level-ttl')
```

# `ttl(db[, opts])`

Adds a `refreshTtl` method to the `db`. When `db.refreshTtl(key)` is called, it will refresh the ttl on the `key`. This adds the ability to make the ttl act like a session manager by calling `refreshTtl()` every time you do `db.get()`.

Also, this respects the locks that [`level-lock`][lock] creates. If `tiny-level-ttl` attempts to delete a key, and the key's write access is locked, it will restart the key's life. (In most cases, this is the desired outcome. If the key is being written to, you would've restarted the key's life anyway. If the key is being deleted, restarting its life will not mess anything up.)

- `db` is a levelup database.
- `opts` is an object with the following properties:
	- `ttl` is a number of milliseconds for how long a key lives in the `db`. Optional; defaults to `3600000`, (1 hour).
	- `checkInterval` is a number of milliseconds for how long the interval between checking keys is. Optional; defaults to `10000`, (10 seconds).
	- `separator` can be a string, buffer, or array of strings/buffers. These represent the separator used in the sub-database. For example, `level-spaces` by default has the separator `~`, so if you store the key `bar`, in the space `foo`, the key will be `~foo~bar`. Defaults to `[ '~', '\xff', '\x00' ]`.

# example

Basic usage:

```js
var level = require('level-mem')
var ttl = require('tiny-level-ttl')

var db = level('/levelmem/does/not/care')
ttl(db, {
	ttl: 1000,
	checkInterval: 50
})

db.put('hi', 'wuzzup') //this sets the ttl

setTimeout(function () { //before key expires
	db.get('hi', function (err, value) {
		console.log(err && err.notFound) // -> null
		console.log(value) // -> 'wazzup'
	})
}, 900)

setTimeout(function () { //after key expires
	db.get('hi', function (err, value) {
		console.log(err && err.notFound) // -> true
		console.log(value) // -> undefined
	})
}, 1100)
```

# install

With [npm](http://nodejs.org/download) do:

```
npm install tiny-level-ttl
```

# license

[VOL](http://veryopenlicense.com/)


[bug-code]: https://gist.github.com/ArtskydJ/65ebbd9cdbcdea9f091e
[levelup]: https://github.com/rvagg/node-levelup
[lock]: https://github.com/substack/level-lock
[spaces]: https://github.com/rvagg/level-spaces
[sublevel]: https://github.com/dominictarr/level-sublevel
[ttl]: https://github.com/rvagg/node-level-ttl
[wiki-ttl]:https://en.wikipedia.org/wiki/Time_to_live


CHANGELOG.md
# 3.1.5

- Fixed tests after `expire-unused-keys` update.

# 3.1.4

- Bumped up the timeout on a test, because it was failing.

# 3.1.3

- Added test for an array of buffer separators. (Fixes [#5](https://github.com/ArtskydJ/tiny-level-ttl/issues/5).)

# 3.1.2

- Added test for buffer separators. (Fixes [#5](https://github.com/ArtskydJ/tiny-level-ttl/issues/5).)

# 3.1.1

- TTL shuts down nicely when the database closes.
- Sped up tests.

# 3.1.0

- Added `separator` option.

# 3.0.0

- Added support for `level-spaces@2`
- The check for whether a key is in a sub level is more thorough.
- Breaking changes: Keys with `'\x00'`, `'\xff'`, or `'~'` in their name are not supported. (In 3.1.0+, they are supported by passing in a `separator` that overrides the defaults.)

# 2.0.3

- Updated `.travis.yml`.

# 2.0.2

- Updated `.gitignore`.

# 2.0.1

- Removed unused dependency.

# 2.0.0

- Added support for `level-spaces@1`.
- Breaking changes: none.

# 1.1.0

- Internally using `level-spaces` instead of `level-sublevel`.

# 1.0.0

- Republished v0.2.0 as v1.0.0

# 0.2.0

- Added tests.

# 0.1.0

- Added `refreshOnGet` option.

# 0.0.3

- Instead of multiple options being passed in, and options object is passed in.
- Added `.travis.yml`.

# 0.0.2

- Added tests.

# 0.0.1

- Initial version.
- Added support for `level-sublevel`.


test.js
var test = require('tap').test
var level = require('level-mem')
var sublevel = require('level-sublevel')
var ttl = require('./index.js')

test('basic functionality', function (t) {
	t.plan(6)
	var db = level('hello')
	db = sublevel(db)
	ttl(db, {ttl: 1000, checkInterval: 50})
	db.put('hi', 'wuzzup')
	setTimeout(function () { //before ttl
		db.get('hi', function (err, value) {
			t.notOk(err, 'did not get an error')
			t.notOk(err && err.notFound, 'did not get a notFound error')
			t.equal(value, 'wuzzup', 'got back the expected value')
		})
	}, 900)
	setTimeout(function () { //after ttl
		db.get('hi', function (err, value) {
			t.ok(err, 'got an error')
			t.ok(err && err.notFound, 'got a notFound error')
			t.notOk(value, 'did not get a value')
			t.end()
		})
	}, 1100)
})

test('delay expiration with put()', function (t) {
	t.plan(10)
	var db = level('hello')
	db = sublevel(db)
	ttl(db, {ttl: 1000, checkInterval: 50})
	db.put('hi', 'wuzzup')
	setTimeout(function () { //delay ttl
		db.get('hi', function (err, value) {
			t.notOk(err, 'did not get an error')
			t.notOk(err && err.notFound, 'did not get a notFound error')
			t.equal(value, 'wuzzup', 'got back the expected value')

			db.put('hi', 'hola', function (err) {
				t.notOk(err, 'did not get an error')
			})
		})
	}, 900)
	setTimeout(function () { //before ttl
		db.get('hi', function (err, value) {
			t.notOk(err, 'did not get an error')
			t.notOk(err && err.notFound, 'did not get a notFound error')
			t.equal(value, 'hola', 'got back the expected value')
		})
	}, 1800)
	setTimeout(function () { //after ttl
		db.get('hi', function (err, value) {
			t.ok(err, 'got an error')
			t.ok(err && err.notFound, 'got a notFound error')
			t.notOk(value, 'did not get a value')
			t.end()
		})
	}, 2000)
})

/*
//levelup does not emit 'get' events, so this probably isn't going to happen very soon

test('delay expiration with get() when refreshOnGet is true', function (t) {
	t.plan(9)
	var db = level('hello')
	db = sublevel(db)
	ttl(db, {ttl: 1000, checkInterval: 50, refreshOnGet: true})
	db.put('hi', 'wuzzup')
	setTimeout(function () { //delay ttl
		db.get('hi', function (err, value) {
			t.notOk(err, 'did not get an error')
			t.notOk(err && err.notFound, 'did not get a notFound error')
			t.equal(value, 'wuzzup', 'got back the expected value')
		})
	}, 900)
	setTimeout(function () { //delay ttl (and test)
		db.get('hi', function (err, value) {
			t.notOk(err, 'did not get an error')
			t.notOk(err && err.notFound, 'did not get a notFound error')
			t.equal(value, 'wuzzup', 'got back the expected value')
		})
	}, 1100)
	setTimeout(function () { //after ttl
		db.get('hi', function (err, value) {
			t.ok(err, 'got an error')
			t.ok(err && err.notFound, 'got a notFound error')
			t.notOk(value, 'did not get a value')
			t.end()
		})
	}, 2300)
})*/

test('still works in a sublevel', function (t) {
	t.plan(10)
	var db = level('hello')
	db = sublevel(db)
	var useDb = db.sublevel('test')
	ttl(useDb, {ttl: 1000, checkInterval: 50})

	var puts = 0
	useDb.on('put', function () { puts++ })

	useDb.put('hi', 'wuzzup')
	db.put('original', 'no_timeout')
	setTimeout(function () { //before ttl
		useDb.get('hi', function (err, value) {
			t.notOk(err, 'did not get an error')
			t.notOk(err && err.notFound, 'did not get a notFound error')
			t.equal(value, 'wuzzup', 'got back the expected value')
		})
	}, 900)
	setTimeout(function () { //after ttl
		useDb.get('hi', function (err, value) {
			t.ok(err, 'got an error')
			t.ok(err && err.notFound, 'got a notFound error')
			t.notOk(value, 'did not get a value')
			t.equal(puts, 1, 'one put() call')

			db.get('original', function (err, value) {
				t.notOk(err, 'original - did not get an error')
				t.notOk(err && err.notFound, 'original - did not get a notFound error')
				t.equal(value, 'no_timeout', 'original - got back the expected value')
				t.end()
			})
		})
		
	}, 1100)
})

test('throw error on bad db', function (t) {
	t.plan(4)
	var db = level('hello')
	try {
		ttl(null)
		t.fail('bad db must throw error')
	} catch (err) {
		t.ok(err.message.indexOf('database')>=0, 'has "database" in error message')
		t.equal(err.message.indexOf('sublevel'), -1, 'does not have "sublevel" in error message')
	}
	try {
		ttl(db)
		t.fail('non subleveled db must throw error')
	} catch (err) {
		t.ok(err.message.indexOf('database')>=0, 'has "database" in error message')
		t.ok(err.message.indexOf('sublevel')>=0, 'has "sublevel" in error message')
	}
	t.end()
})

test('no errors if entry is deleted', function (t) {
	var db = level('hello')
	db = sublevel(db)
	ttl(db, {ttl: 1000, checkInterval: 100})
	db.on('error', t.fail.bind(t))
	db.put('hi', 'wuzzup', t.notOk.bind(t))
	setTimeout(db.del.bind(db, 'hi', t.notOk.bind(t)), 200) //before ttl
	setTimeout(t.end.bind(t), 1300) //if an error is thrown, it should show
})


test/testLocking.js
var test = require('tape')
var level = require('level-mem')
var ttl = require('../index.js')
var lock = require('level-lock')

test('basic locking', function (t) {
	var db = level('hello')
	ttl(db, {
		ttl: 100,
		checkInterval: 20
	})
	db.put('hi', 'wuzzup')
	var unlock = undefined
	db.on('error', t.fail.bind(t))
	setTimeout(function () { // before ttl
		unlock = lock(db, 'hi', 'w')
		t.ok(unlock, 'got lock')
	}, 70)
	setTimeout(function () { // after ttl
		unlock()
		db.get('hi', function (err, value) {
			t.notOk(err, 'no error')
			t.notOk(err && err.notFound, 'no "notFound" error')
			t.ok(value, 'key exists')
		})
	}, 130)
	setTimeout(function () {
		db.get('hi', function (err, value) { // does it retry?
			t.ok(err, 'error')
			t.ok(err && err.notFound, '"notFound" error')
			t.notOk(value, 'key does not exist')
		})
		t.end()
	}, 250)
})

test('more locking and deletions', function (t) {
	var db = level('hello')
	ttl(db, {
		ttl: 100,
		checkInterval: 20
	})
	db.put('hi', 'wuzzup')
	var unlock = undefined
	db.on('error', t.fail.bind(t))
	setTimeout(function () { // before ttl
		unlock = lock(db, 'hi', 'w')
		t.ok(unlock, 'got lock')
	}, 70)
	setTimeout(function () { // after ttl
		db.del('hi', function (err) {
			t.notOk(err, 'no error on deletion')
			unlock()
		})
	}, 130)
	setTimeout(t.end.bind(t), 250)
})


test/testShutdown.js
var test = require('tape')
var level = require('level-mem')
var ttl = require('../index.js')

test('closes handlers on closing event', function (t) {
	var db = new level()
	db.sublevel = level
	ttl(db)

	t.equal(typeof db.refreshTtl, 'function', 'not shutdown yet')
	db.emit('closing')
	t.equal(typeof db.refreshTtl, 'undefined', 'is shutdown')

	var fail = t.fail.bind(t, 'uncaught exception')
	process.on('uncaughtException', fail)
	setTimeout(function () {
		process.removeListener('uncaughtException', fail)
		t.end()
	}, 20)
})


index.js
var Expirer = require('expire-unused-keys')
var xtend = require('xtend')
var lock = require('level-lock')
var spaces = require('level-spaces')
var defaultOpts = {
	ttl: 3600000,
	checkInterval: 10000,
	separator: [ '~', '\xff', '\x00' ]
}

function isFromThisSpace(key, separator) {
	if (!Array.isArray(separator)) {
		separator = [ separator ]
	}
	return separator.every(function (sep) {
		return key.indexOf(sep) === -1
	})
}

function createResponses(expirer, separator) {
	var responses = {}
	responses.del = function(key) {
		if (isFromThisSpace(key, separator)) {
			expirer.forget(key)
		}
	}
	responses.put = function(key) {
		if (isFromThisSpace(key, separator)) {
			expirer.touch(key)
		}
	}
	responses.batch = function(cmds) {
		cmds.forEach(function(cmd) {
			responses[cmd.type](cmd.key)
		})
	}
	return responses
}

function makeChildDb(db, separator) {
	return db.sublevel ?
		db.sublevel('expirer') :
		spaces(db, 'expirer', { separator: separator })
}

module.exports = function ttl(db, opts) {
	if (!db) {
		throw new Error('You must pass a level database to ttl()')
	}
	var options = xtend(defaultOpts, opts)
	var separator = options.separator
	var childDb = makeChildDb(db, separator)
	var expirer = new Expirer(options.ttl, childDb, options.checkInterval)
	var responses = createResponses(expirer, separator)

	db.on('put', responses.put)
	db.on('del', responses.del)
	db.on('batch', responses.batch)
	db.once('closing', shutdown)

	expirer.on('expire', function (key) {
		var unlock = lock(db, key, 'w')
		if (unlock) {
			db.del(key, unlock)
		} else {
			expirer.touch(key) // Retry later
		}
	})
	db.refreshTtl = expirer.touch

	function shutdown() {
		delete db.refreshTtl
		db.removeListener('put', responses.put)
		db.removeListener('del', responses.del)
		db.removeListener('batch', responses.batch)
		expirer.stop()
	}
}
